plugins/woocommerce-admin/client/customize-store/index.tsx

```diff
@@ -0,0 +1,193 @@
+
+export type customizeStoreStateMachineEvents =
+	| introEvents
+	| designWithAiEvents
+	| assemblerHubEvents;
+
+export const customizeStoreStateMachineServices = {
+	...introServices,
+};
+
+export const customizeStoreStateMachineActions = {
+	...introActions,
+};
+export const customizeStoreStateMachineDefinition = createMachine( {
+	id: 'customizeStore',
+	initial: 'intro',
+	predictableActionArguments: true,
+	preserveActionOrder: true,
+	schema: {
+		context: {} as customizeStoreStateMachineContext,
+		events: {} as customizeStoreStateMachineEvents,
+		services: {} as {
+			fetchThemeCards: { data: ThemeCard[] };
+		},
+	},
+	context: {
+		intro: {
+			themeCards: [] as ThemeCard[],
+			activeTheme: '',
+		},
+	} as customizeStoreStateMachineContext,
+	states: {
+		intro: {
+			id: 'intro',
+			initial: 'preIntro',
+			states: {
+				preIntro: {
+					invoke: {
+						src: 'fetchThemeCards',
+						onDone: {
+							target: 'intro',
+							actions: [ 'assignThemeCards' ],
+						},
+					},
+				},
+				intro: {
+					meta: {
+						component: Intro,
+					},
+				},
+			},
+			on: {
+				DESIGN_WITH_AI: {
+					target: 'designWithAi',
+				},
+				SELECTED_ACTIVE_THEME: {
+					target: 'assemblerHub',
+				},
+				CLICKED_ON_BREADCRUMB: {
+					target: 'backToHomescreen',
+				},
+				SELECTED_NEW_THEME: {
+					target: '? Appearance Task ?',
+				},
+				SELECTED_BROWSE_ALL_THEMES: {
+					target: '? Appearance Task ?',
+				},
+			},
+		},
+		designWithAi: {
+			initial: 'preDesignWithAi',
+			states: {
+				preDesignWithAi: {
+					always: {
+						target: 'designWithAi',
+					},
+				},
+				designWithAi: {
+					meta: {
+						component: DesignWithAi,
+					},
+				},
+			},
+			on: {
+				THEME_SUGGESTED: {
+					target: 'assemblerHub',
+				},
+			},
+		},
+		assemblerHub: {
+			on: {
+				FINISH_CUSTOMIZATION: {
+					target: 'backToHomescreen',
+				},
+			},
+		},
+		backToHomescreen: {},
+		'? Appearance Task ?': {},
+	},
+} );
+
+export const CustomizeStoreController = ( {
+	actionOverrides,
+	servicesOverrides,
+}: {
+	actionOverrides: Partial< typeof customizeStoreStateMachineActions >;
+	servicesOverrides: Partial< typeof customizeStoreStateMachineServices >;
+} ) => {
+	useFullScreen( [ 'woocommerce-customize-store' ] );
+
+	const augmentedStateMachine = useMemo( () => {
+		return customizeStoreStateMachineDefinition.withConfig( {
+			services: {
+				...customizeStoreStateMachineServices,
+				...servicesOverrides,
+			},
+			actions: {
+				...customizeStoreStateMachineActions,
+				...actionOverrides,
+			},
+			guards: {},
+		} );
+	}, [ actionOverrides, servicesOverrides ] );
+
+	const [ state, send, service ] = useMachine( augmentedStateMachine, {
+		devTools: process.env.NODE_ENV === 'development',
+	} );
+
+	// eslint-disable-next-line react-hooks/exhaustive-deps -- false positive due to function name match, this isn't from react std lib
+	const currentNodeMeta = useSelector( service, ( currentState ) =>
```

Comment:
These lines seem to be duplicates of the core-profiler. Do you think we might optimize it by creating more reusable components/utils in the future?

Reply:
I've refactored out the ones I was certain about in https://github.com/woocommerce/woocommerce/pull/39581 , so the remaining duplicates are the ones that I'd like to see more evolution before deciding on how to generalise